### PR TITLE
Abstract integrator

### PR DESCRIPTION
**Pull request by thomaspinder**
## Type of changes

- [ ] Bug fix
- [X] New feature
- [ ] Documentation / docstrings
- [ ] Tests
- [ ] Other

## Checklist

- [X] I've formatted the new code by running `poetry run pre-commit run --all-files --show-diff-on-failure` before committing.
- [X] I've added tests for new code.
- [x] I've added docstrings for the new code.

## Description

This PR enables users to define their own integration strategy for computing the expected log likelihood under the variational distribution.

**Note:** A lot of docs have been changed here as `docs/api` had been committed somehwere along the way. The key files that have been changed are `gpjax/likelihoods.py` and `gpjax/integrators.py`.

Issue Number: #270 #282

### CODE REVIEW

File: gpjax/integrators.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+@dataclass
+class GHQuadratureIntegrator(AbstractIntegrator):
+    num_points: int = 20
+
+    def integrate(
+        self,
+        fun: Callable,
+        y: Float[Array, "N D"],
+        mean: Float[Array, "N D"],
+        sigma2: Float[Array, "N D"],
```
Comment (daniel-dodd) [inline gpjax/integrators.py:50]: I would be explicit with `sigma2`, renaming to `variance`? Or explain in a docstring.

File: gpjax/integrators.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+        **likelihood_params: Any,
+    ) -> Float[Array, " N"]:
+        gh_points, gh_weights = np.polynomial.hermite.hermgauss(self.num_points)
+        sd = jnp.sqrt(sigma2)
```
Comment (daniel-dodd) [inline gpjax/integrators.py:54]: This might be a potential numerical issue to keep an eye on. If sigma2 gets sufficiently small, the gradients will explode.

File: gpjax/integrators.py [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+@dataclass
+class AnalyticalGaussianIntegrator(AbstractIntegrator):
+    def integrate(
+        self,
+        fun: Callable,
+        y: Float[Array, "N D"],
+        mean: Float[Array, "N D"],
+        sigma2: Float[Array, "N D"],
+        **likelihood_params: Any,
```
Comment (daniel-dodd) [inline gpjax/integrators.py:69]: Instead of `likelihood_params` dictionary, why not just define integrator on the likelihood akin to the kernel compute engine? You could have:

```python
@dataclass
AbstractIntegrator:
    likelihood: AbstractLikelihood

```

File: gpjax/integrators.py [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+from jaxtyping import Float
+import numpy as np
+from simple_pytree import Pytree
+
+from gpjax.typing import Array
+
+
+@dataclass
+class AbstractIntegrator(Pytree):
+    @abstractmethod
```
Comment (daniel-dodd) [inline gpjax/integrators.py:18]: This does not need to be a Pytree.

File: gpjax/likelihoods.py [hunk 5]
```diff
@@ -85,6 +92,15 @@ def link_function(self, f: Float[Array, "..."]) -> tfd.Distribution:
...
 
+    def expected_log_likelihood(
+        self,
+        y: Float[Array, "N D"],
+        mu: Float[Array, "N D"],
+        sigma2: Float[Array, "N D"],
+    ):
+        log_prob = vmap(lambda f, y: self.link_function(f).log_prob(y))
+        return self.integrator(fun=log_prob, y=y, mean=mu, sigma2=sigma2, **self.dict())
+
```
Comment (daniel-dodd) [inline gpjax/likelihoods.py:103]: This is nice. Maybe I would either call this a variational expectation --- or reference this terminology in a docstring --- in keeping with the literature.

File: gpjax/base/module.py [hunk 6]
```diff
@@ -244,6 +244,9 @@ def _get_trainables(meta_leaf):
 
         return meta_map(_get_trainables, self)
 
+    def dict(self):
+        return {k: v for k, v in dataclasses.asdict(self).items()}
+
```
Comment (daniel-dodd) [inline gpjax/base/module.py:249]: I am strongly against having this.

**Review by daniel-dodd — COMMENTED** [review]
Hey @thomaspinder, looking good. Like the structure of it. Just left a few thoughts and suggestions.

File: gpjax/integrators.py (hunk 3 above)
Comment (daniel-dodd) [reply gpjax/integrators.py:69]: Bad suggestion. Does not allow the class to be initialised before supplying it to the likelihood.

**Review by daniel-dodd — COMMENTED** [review]
Looks good @thomaspinder, just minor comment about docstrings, you need `likelihood` instead of `**likelihood_params` in each integrator class.
